### ISSUE TITLE
Add specialized builtins for handling Value?

### ISSUE THREAD

**Issue by effectfully** [IntersectMBO/plutus#6585]
@colll78 [reports](https://x.com/phil_uplc/status/1845830624214466773)

> lack of specialized builtins for values (`builtinUnionValue`, `builtinValueOf`, etc)

I'm not sure what the best way of doing this is though. Do we add `Value` as a built-in type plus some conversions to `ByteStringMap` or `Data`? Can `Value` change in future? In that case hardcoding the details of a specific implementation in the type signature (i.e. returning a `ByteStringMap`) of a conversion builtin is perhaps not wise, while returning `Data` is maybe fine. 

Or do we just alias `Value = Data` in builtins?

Or maybe we can get decent performance without specialized builtins once we have `ByteStringMap` and some general builtins over it? I can only think of one way of doing that though, but it's not pretty: #6587.

**Comment by rvcas** [comment]
adding something named Value as a builtin type would be mildly annoying at least cause naming conflicts with Cek Value. I think ByteStringMap is the way for this, bringing a standard library concept down to UPLC does not seem like a good idea.

**Comment by effectfully** [comment]
> adding something named Value as a builtin type would be mildly annoying at least cause naming conflicts with Cek Value.

Personally I think this name clash is insignificant, particularly given that it's `CekValue`, not just `Value`. Anyway, doesn't really matter and I'd be happy with a different name.

> I think ByteStringMap is the way for this, bringing a standard library concept down to UPLC does not seem like a good idea.

The issue here is that `unionValue` seems to be a popular function and it's really hard to express it as a built-in without some heavy artillery ([1](https://github.com/IntersectMBO/plutus/issues/6587), [2](https://github.com/IntersectMBO/plutus/issues/6563)). And as a non-builtin it's quite expensive. Or do you think we can afford having `unionValue` to be inefficient?

> bringing a standard library concept down to UPLC does not seem like a good idea

Perhaps. @colll78 you have a different view, don't you?

**Comment by colll78** [comment]
> bringing a standard library concept down to UPLC does not seem like a good idea.

This does not make sense to me. Why do you consider Value as BuiltinData to be a “standard library concept”? Maybe you have misunderstood. My proposal is to add builtin functions for working with `AsBuiltinData Value`.  

UPLC is not a general-purpose programming language. It is a language specifically designed to write validation logic to set constraints for the creation and transfer of Value and Data across the Cardano blockchain. Given that half of the entire purpose of this language is to add constraints to the creation and transfer of Value, why should Value be a “standard library concept”? 

The whole value proposition behind domain specific languages (ie UPLC) is that the design is informed by and tailored to, the problem space. As such, it should be better suited to the task in that domain than a general-purpose language. To me it seems like this has not been taken into consideration much historically in the design of Plutus. How is there not first-class language support for Values in a domain specific language designed primarily to manage the transfer of Value???

**Comment by colll78** [comment]
Also BuiltinByteStringMap is not sufficient for many production use-cases, because you will have to first coerce the relevant data in the script context from BuiltinData to `BuiltinList (CS (BuiltinList (TN, Amount)))` and then convert it to `BuiltinByteStringMap` and then operate on it and then convert the inner map to `BuiltinByteStringMap`. 

This wouldn’t be the case if we could support multiple variants of the ScriptContext (legacy with only original BuiltinData types, and modern with the new types ie BuiltinArray and map). But this is a long long ways away.

**Comment by mpetruska** [comment]
IMHO It makes sense to a dedicated `Value` builtin type, as it carries some additional constraints on top of just the type of the keys (e.g. the Int values are always non-negative). Also it seems that plenty of validator scripts are frequently looping over them just for checking the existence of an nft, which can be expensive in some (although maybe rare) cases.
Another pro for it could be security. If some validator authors forget to cover the "token dust attack" (for example), the builtin operations might be implemented in a way to add automatic checks for it.

**Comment by mpetruska** [comment]
With regards to the naming/"standard library concept" issue:
- the word `value` is extremely overused, since it is a general programming concept (as in: "all values have types")
- usually apis, libraries and frameworks directly expose types/functions that represent their core concepts

**Comment by keyan-m** [comment]
This is a nice initiation, and I think @colll78 articulated quite nicely. There are so many utility functions I end up writing for almost all non-trivial contracts (the one I can think of right away, is a function for validating a UTxO is "authentic," i.e. carries a specific NFT). While I understand adding features to a language should be done conservatively, this kind of repetition is bloating the chain.

**Comment by rvcas** [comment]
I still just think more functions for working with PlutusData is more appropriate. The concept of Value (Multiasset thingy) doesn't exist in the VM, it's just PlutusData. But I don't feel too strongly, I just think focusing on more performant primitives is more worth it than embedding high level constructs into the machine.

The notion that Plutus VM is general purpose is silly. I was certainly not advocating for some vague notion of general purpose.

**Comment by colll78** [comment]
> I still just think more functions for working with PlutusData is more appropriate. The concept of Value (Multiasset thingy) doesn't exist in the VM, it's just PlutusData. But I don't feel too strongly, I just think focusing on more performant primitives is more worth it than embedding high level constructs into the machine.

I agree. I am suggesting exactly that, functions for dealing `BuiltinData` that is a map of inner maps where the keys are integer quantities (ie `BuiltinData` that corresponds to Value. I don't see `Value` as a high-level construct, it is a primitive of the domain space.

**Comment by Unisay** [comment]
`Value` is now a builtin type with dedicated builtins, so I'm closing this.

[CIP-0153](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0153) added the `BuiltinValue` type with `insertCoin`, `lookupCoin`, `unionValue`, `valueContains`, `valueData`, `unValueData`, `scaleValue` — live on mainnet since protocol version 11.

[CIP-0168](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0168) (merged 2026-08-18) adds `policies` (already in plutus 1.67.0.0, see #7858), `keepPolicies`, `dropPolicies`, and `assetCount`, targeted at Plutus V4.
